### PR TITLE
Best.Conventional.Roslyn upgrade

### PR DESCRIPTION
**Pull request by murlakatam**
- Upgrading target framework to latest LTS version
- Updating dependencies, making sure dependencies are not strict with minor versions and support wide range of variations on consumer side
- Including Conventional.Roslyn.Analyzers.dll as dependency in output nuget/lib/net6.0/ so that writing custom SolutionDiagnosticAnalyzerConventionSpecification is possible 
 as DiagnosticResult type is missing from public contract currently
 
image

### CODE REVIEW

**Comment by murlakatam** [comment]
@andrewabest FYI

**Comment by andrewabest** [comment]
Thanks @murlakatam - I'll try and take a look at this shortly.

File: src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj [hunk 1]
```diff
@@ -5,8 +5,8 @@
     </PropertyGroup>
 
     <ItemGroup>
-      <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="3.6.0" />
-      <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.6.0" />
+      <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="[4.0.1,)" />
```
Comment (andrewabest) [inline src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj:8]: I'd prefer to use fixed versions for dependencies - it is important (to me) to be able to easily understand what versions of dependencies are in what build of the library. 

A solution I _should_ put in place for this is to set up dependabot and let it do its thing. I'll add that to my TODO list. For now, could you please change these versions to an appropriate fixed version?

File: src/Roslyn/Conventional.Roslyn/Conventional.Roslyn.csproj [hunk 2]
```diff
@@ -17,10 +17,10 @@
     </PropertyGroup>
 
     <ItemGroup>
-      <PackageReference Include="Best.Conventional" Version="9.0.3" />
-      <PackageReference Include="Microsoft.Build.Locator" Version="1.2.6" />
-      <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.6.0" />
-      <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.MSBuild" Version="3.6.0" />
+      <PackageReference Include="Best.Conventional" Version="[9.0.3,)" />
```
Comment (andrewabest) [inline src/Roslyn/Conventional.Roslyn/Conventional.Roslyn.csproj:20]: As above, if you can specify fixed versions that would be good.

File: src/Roslyn/Conventional.Roslyn/Conventional.Roslyn.csproj [hunk 3]
```diff
@@ -30,6 +30,21 @@
     <ItemGroup>
         <None Update="tools\*.ps1" CopyToOutputDirectory="Always" Pack="true" PackagePath="tools" />
         <None Include="$(OutputPath)\Conventional.Roslyn.Analyzers.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
+        
     </ItemGroup>
 
+    <PropertyGroup>
```
Comment (andrewabest) [inline src/Roslyn/Conventional.Roslyn/Conventional.Roslyn.csproj:36]: I took a dive down the rabbit hole to find out why this required so much msbuild chicanery to work.

https://github.com/dotnet/sdk/issues/6688

It is a bit of a sad state that you need to go to these lengths to package a project dependency in a nuget package 😢

**Review by andrewabest — CHANGES_REQUESTED** [review]
Thanks for your efforts here @murlakatam

I've requested changes to convert version ranges to fixed versions, but once that is done, this should be good to ship!

File: src/Roslyn/Conventional.Roslyn/Conventional.Roslyn.csproj (hunk 3 above)
Comment (murlakatam) [reply src/Roslyn/Conventional.Roslyn/Conventional.Roslyn.csproj:36]: Yeah, I've found that workaround some time ago on one of my projects, quite sad to see this is still the case.

File: src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj (hunk 1 above)
Comment (murlakatam) [reply src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj]: hopefully these particular versions are good enough

File: src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj [hunk 4]
```diff
@@ -5,8 +5,8 @@
     </PropertyGroup>
 
     <ItemGroup>
-      <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="3.6.0" />
-      <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.6.0" />
+      <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.5.0" />
```
Comment (andrewabest) [inline src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj:8]: @murlakatam should we go with `4.7.0` here, given that is the latest version and what we are using in the test project, or do you need `4.5.0` for a specific reason?

Comment (murlakatam) [reply src/Roslyn/Conventional.Roslyn.Analyzers/Conventional.Roslyn.Analyzers.csproj]: @andrewabest good point, I was constraint by another dependency of mine that was strictly wired to 4.5.0, but that's my pain. I will workaround that. Pushed a commit with 4.7.0 change

**Review by andrewabest — APPROVED** [review]
Thanks for accommodating the changes @murlakatam!

**Comment by murlakatam** [comment]
> Thanks for accommodating the changes @murlakatam!

No worries, @andrewabest. Love your work. 
Question: seems like the merge only pushed the main nuget Best.Conventional v11.0.1 
what about the Best.Conventional.Roslyn nuget (I think it still sits at version 8.0.4) 
image

**Comment by andrewabest** [comment]
> > Thanks for accommodating the changes @murlakatam!
> 
> No worries, @andrewabest. Love your work. Question: seems like the merge only pushed the main nuget Best.Conventional v11.0.1 what about the Best.Conventional.Roslyn nuget (I think it still sits at version 8.0.4) 

So the update to dotnet 6.0 has surfaced a bug in the CI build.

Repro:

```
dotnet pack "{pathToRoot}\conventional\src\Roslyn\Conventional.Roslyn\Conventional.Roslyn.csproj" --configuration Release --no-build --verbosity d
```

Will hit the error
`C:\Program Files\dotnet\sdk\7.0.403\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(175,5): error NETSDK1085: The 'NoBuild' property was set to true but the 'Build' target was invoked.`

I'm sorting it out now, hopefully will have a build soonish.

**Comment by andrewabest** [comment]
@murlakatam all fixed, it is now published as `9.0.5`. I've also updated the Readme with a build status badge / nuget status badge for the roslyn project to make it a bit easier to know when a new version is published.

**Comment by murlakatam** [comment]
> @murlakatam all fixed, it is now published as `9.0.5`. I've also updated the Readme with a build status badge / nuget status badge for the roslyn project to make it a bit easier to know when a new version is published.

Thank you.

**Pull request merged**